File: src/python/titanic.py
import sys
import pandas as pd
import os
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report


# === 13. Load train.csv ===
print("\nLoading train.csv ...")
train = pd.read_csv("/data/train.csv")
print("Train loaded. Shape:", train.shape)
print("Train columns:", list(train.columns))

# === 14. Explore / clean the data ===
print("\nExploring and cleaning train data ...")
print("Missing values per column (train):\n", train.isnull().sum())

# Drop irrelevant Cabin column
print("Dropping column: Cabin (if exists)")
train = train.drop(["Cabin"], axis=1)
print("Columns after dropping Cabin:", list(train.columns))

# Fill missing values
age_na_before = train["Age"].isna().sum()
train["Age"].fillna(train["Age"].median(), inplace=True)
print("Filled Age missing with median. Before:", age_na_before, "After:", train["Age"].isna().sum())

emb_na_before = train["Embarked"].isna().sum()
train["Embarked"].fillna(train["Embarked"].mode()[0], inplace=True)
print("Filled Embarked missing with mode. Before:", emb_na_before, "After:", train["Embarked"].isna().sum())

# Encode categorical data
print("Encoding Sex as {male:0, female:1}")
train["Sex"] = train["Sex"].map({"male": 0, "female": 1})
print("One-hot on Embarked with drop_first=True")
train = pd.get_dummies(train, columns=["Embarked"], drop_first=True)
print("Columns after get_dummies (train):", list(train.columns))

print("Train cleaned. New shape:", train.shape)

# === 15. logistic regression model to predict survivability ===
features = ["Pclass", "Sex", "Age", "SibSp", "Parch", "Fare",
            'Embarked_Q', 'Embarked_S']
print("\nUsing features:", features)
X = train[features]
y = train["Survived"]
print("X shape (train):", X.shape, " y shape:", y.shape)

scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)
print("Scaled X (train) shape:", X_scaled.shape)

model = LogisticRegression(max_iter=1000, random_state=42)
model.fit(X_scaled, y)
print("Model training complete.")

# === 16. Measure the accuracy ===
y_pred = model.predict(X_scaled)
acc = accuracy_score(y, y_pred)
print("Training accuracy:", f"{acc:.4f}")
print("Confusion matrix (train):\n", confusion_matrix(y, y_pred))

# === 17–18. Load test.csv, predict, and measure accuracy ===
print("\nLoading test.csv ...")
test = pd.read_csv("/data/test.csv")
print("Test loaded. Shape:", test.shape)
print("Missing values per column (test):\n", test.isnull().sum())

print("Filling test Age and Fare missing with medians")
test["Age"].fillna(test["Age"].median(), inplace=True)
test["Fare"].fillna(test["Fare"].median(), inplace=True)

print("Encoding train Sex/Embarked (as written above in your script)")
test["Sex"] = test["Sex"].map({"male": 0, "female": 1})
test = pd.get_dummies(test, columns=["Embarked"], drop_first=True)
print("Columns after get_dummies (test):", list(test.columns))

print("Selecting test features with same columns/order as train")
X_test = test[features]
print("X_test shape:", X_test.shape)

print("Scaling test with train scaler and predicting")
X_test_scaled = scaler.transform(X_test)
y_test_pred = model.predict(X_test_scaled)
print("First 10 test predictions:", y_test_pred[:10])

print("Loading gender_submission.csv to compute test accuracy ...")
gender = pd.read_csv("/data/gender_submission.csv")
print("gender_submission shape:", gender.shape)

pred = pd.DataFrame({
    "PassengerId": test["PassengerId"],
    "Predicted": y_test_pred
})

merged = pred.merge(gender, on="PassengerId", how="inner")
print("Merged data shape:", merged.shape)

agreement = (merged["Predicted"] == merged["Survived"]).mean()
print(f"Agreement with gender_submission.csv(testing accuracy): {agreement:.4f}")

File: src/r/titanic.R
suppressPackageStartupMessages({
  library(readr)
  library(dplyr)
  library(fastDummies)
})

DATA_DIR <- "/data"

# ---------- helpers ----------
print_missing <- function(df, title) {
  cat(paste0("\nMissing values per column (", title, "):\n"))
  print(sapply(df, function(x) sum(is.na(x) | (is.character(x) & x == ""))))
}

scale_fit <- function(df_num) {
  mu  <- sapply(df_num, mean, na.rm = TRUE)
  sdv <- sapply(df_num, sd,   na.rm = TRUE)
  sdv[sdv == 0] <- 1
  list(mu = mu, sd = sdv)
}
scale_apply <- function(df, stats) {
  out <- df
  for (nm in names(stats$mu)) {
    out[[nm]] <- (out[[nm]] - stats$mu[[nm]]) / stats$sd[[nm]]
  }
  out
}

# 清洗函数（train / test 共用）
clean_titanic <- function(df, is_train = TRUE) {
  # 1) 丢弃 Cabin（若存在）
  if ("Cabin" %in% names(df)) {
    df <- df %>% select(-Cabin)
    cat("Dropped column: Cabin\n")
  }

  # 2) 填充 Age / Fare / Embarked
  if ("Age" %in% names(df)) {
    before <- sum(is.na(df$Age))
    med    <- median(df$Age, na.rm = TRUE)
    df     <- df %>% mutate(Age = ifelse(is.na(Age), med, Age))
    cat("Filled Age median. Before:", before, " After:", sum(is.na(df$Age)), "\n")
  }
  if ("Fare" %in% names(df)) {
    before <- sum(is.na(df$Fare))
    med    <- median(df$Fare, na.rm = TRUE)
    df     <- df %>% mutate(Fare = ifelse(is.na(Fare), med, Fare))
    cat("Filled Fare median. Before:", before, " After:", sum(is.na(df$Fare)), "\n")
  }
  if ("Embarked" %in% names(df)) {
    before <- sum(is.na(df$Embarked) | df$Embarked == "")
    df     <- df %>%
      mutate(Embarked = ifelse(is.na(Embarked) | Embarked == "", "S", Embarked))
    cat("Filled Embarked='S'. Before:", before,
        " After:", sum(is.na(df$Embarked) | df$Embarked == ""), "\n")
  }

  # 3) Sex 数值化
  if ("Sex" %in% names(df)) {
    df <- df %>% mutate(Sex = ifelse(Sex == "male", 0L, 1L))
  }

  # 4) One-hot for Embarked（以 S 为基准）
  if ("Embarked" %in% names(df)) {
    df <- df %>%
      mutate(Embarked = factor(Embarked, levels = c("S", "C", "Q"))) %>%
      fastDummies::dummy_cols(
        select_columns = "Embarked",
        remove_selected_columns = TRUE,
        remove_first_dummy = TRUE # 删除 S，保留 Embarked_C / Embarked_Q
      )
  }

  df
}

# ========== 13. Load train.csv ==========
cat("\nLoading train.csv ...\n")
train <- readr::read_csv(file.path(DATA_DIR, "train.csv"), show_col_types = FALSE)
cat("Train shape:", paste(dim(train), collapse = " x"), "\n")
cat("Train columns:", paste(names(train), collapse = ", "), "\n")
print_missing(train, "train")

# ========== 14. Explore / clean ==========
cat("\nExploring / cleaning train ...\n")
train <- clean_titanic(train, is_train = TRUE)
cat("\nData cleaned! New shape:", paste(dim(train), collapse = " x"), "\n")
cat("Columns:", paste(names(train), collapse = ", "), "\n")

# ========== 15. Logistic regression on training set ==========
features <- c("Pclass","Sex","Age","SibSp","Parch","Fare","Embarked_C","Embarked_Q")
cat("\nUsing features:", paste(features, collapse = ", "), "\n")

# 保证 dummy 列存在
for (nm in c("Embarked_C","Embarked_Q")) {
  if (!nm %in% names(train)) train[[nm]] <- 0L
}

X <- train[, features, drop = FALSE]
y <- train$Survived


num_cols <- names(X)[sapply(X, is.numeric)]
stats    <- scale_fit(X[, num_cols, drop = FALSE])
X_scaled <- X
X_scaled[, num_cols] <- scale_apply(X[, num_cols, drop = FALSE], stats)

fit <- glm(y ~ ., data = cbind(y = y, X_scaled), family = binomial())
cat("Model training complete.\n")

# ========== 16. Training accuracy ==========
p_train <- ifelse(predict(fit, type = "response") > 0.5, 1L, 0L)
train_acc <- mean(p_train == y)
cat("Training accuracy:", sprintf("%.4f", train_acc), "\n")

# ========== 17–18. Test: load / predict / agreement ==========
cat("\nLoading test.csv ...\n")
test <- readr::read_csv(file.path(DATA_DIR, "test.csv"), show_col_types = FALSE)
cat("Test shape:", paste(dim(test), collapse = " x"), "\n")
print_missing(test, "test")

cat("\nCleaning test ...\n")
test <- clean_titanic(test, is_train = FALSE)

# 对齐列（确保 Embarked_C / Embarked_Q 存在）
for (nm in c("Embarked_C","Embarked_Q")) {
  if (!nm %in% names(test)) test[[nm]] <- 0L
}

X_test <- test[, features, drop = FALSE]
X_test_scaled <- X_test
X_test_scaled[, num_cols] <- scale_apply(X_test[, num_cols, drop = FALSE], stats)

p_test <- ifelse(predict(fit, newdata = X_test_scaled, type = "response") > 0.5, 1L, 0L)
cat("First 10 test predictions:", paste(p_test[1:min(10, length(p_test))], collapse = " "), "\n")

# Agreement vs gender_submission.csv
gender_path <- file.path(DATA_DIR, "gender_submission.csv")
if (file.exists(gender_path)) {
  gender <- readr::read_csv(gender_path, show_col_types = FALSE)
  merged <- merge(
    data.frame(PassengerId = test$PassengerId, Predicted = p_test),
    gender, by = "PassengerId", all = FALSE
  )
  cat("Merged rows:", nrow(merged), "\n")
  agreement <- mean(merged$Predicted == merged$Survived)
  cat("Agreement with gender_submission.csv:", sprintf("%.4f", agreement), "\n")
} else {
  cat("gender_submission.csv not found; skipping agreement metric.\n")
}
